Handle sum from -∞ to k in eval_sum_residue

t-sibiraj opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open.

old OP notes:

Implemented a new function `get_function_symmetry` to determine whether a rational function is even, odd, or neither based on the symmetry of the numerator and denominator.

Added logic to handle summation of rational functions(where the difference between the degree of the denominator and numerator is >= 2) for the interval (-infinity, finite number) based on symmetry inside the `eval_sum_residue`
 -->
Fixes #27824  

#### Brief description of what is fixed or changed
Summations involving (-oo,k) limits can now be handled by `eval_sum_residue`.

#### Other comments
Although limits were checked for being Integer or Infinite, the case of -oo, Integer was not being handled.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* concrete
   * Added logic for handling summations of rational functions over the interval [-∞, k]
<!-- END RELEASE NOTES -->

t-sibiraj committed:
Add summation handling for interval (-infinity to finite number)

t-sibiraj committed:
author: add Thangaraju Sibiraj to .mailmap

t-sibiraj committed:
using substitution instead of calling eval_sum_residue twice when calculating summation over -infinity to finite number

t-sibiraj committed:
fix code to pass code quality

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1536,6 +1536,24 @@ def get_residue_factor(numer, denom, alternating):
+            # Apply the decomposition formula:
+            # (-oo, a) = (-oo, +oo) - (a+1, +oo)
+            # (-oo, a) = (a, +oo) for even function
+            # (-oo, a) = -(a, +oo) for odd function
+            flipped_f = f.subs(i, -i)
+            is_even = True if flipped_f == f else False
+            res = eval_sum_residue(flipped_f, (i, b, S.Infinity))
```
I didn't mean to do the substitution literally. The code below checks that the function is even doesn't it? Does that not mean that we can just swap the limits?

t-sibiraj replied:
Hi @oscarbenjamin I don't understand what you mean by **_not substituting literally_** could you please clarify ? and also doesn't swapping the limits trigger the following if

https://github.com/sympy/sympy/blob/064958283f4e077bb77c9bdd79cb5c3285575d51/sympy/concrete/summations.py#L1506-L1508

t-sibiraj replied:
Hi @oscarbenjamin I don't understand what you mean by **not substituting literally** could you please clarify ? and also doesn't swapping the limits trigger the following if
https://github.com/sympy/sympy/blob/064958283f4e077bb77c9bdd79cb5c3285575d51/sympy/concrete/summations.py#L1506-L1508

oscarbenjamin replied:
The code below checks that the function is even. If the function is even then we don't need to do `.subs(i, -i)`.

oscarbenjamin commented:
I'm talking about this code that already checks for an even function:
https://github.com/sympy/sympy/blob/df0ad48a7695ac57345d0390bb3e3440b135d517/sympy/concrete/summations.py#L1559
We don't need to add any new check for an even function. We just need to change the code so that when it is an even function if the limits are `(-oo, a)` then they get flipped to `(-a, oo)`.

t-sibiraj commented:
Okie got it I understand it now thank you so much

t-sibiraj committed:
I moved the code that handles the even and odd functions into the already existing function that manages that logic

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1535,10 +1535,14 @@ def get_residue_factor(numer, denom, alternating):
-    if not (a.is_finite and b is S.Infinity):
-        return None
-
     if not is_even_function(numer, denom):
+        #for odd function flip the limit and negate
+        #limit (-oo, a) is flipped to (-a, oo) and the answer is negated
+        if a is S.NegativeInfinity and b.is_finite:
+            res = eval_sum_residue(f, (i, -b, S.Infinity))
```
We don't know that it is an odd function.

t-sibiraj committed:
Add function to classify rational function symmetry as even, odd, or neither

t-sibiraj committed:
fix code to pass all the test cases

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1584,6 +1584,16 @@ def test_summation_by_residues():
+    assert summation((k), (k, -oo, 0)) == oo
+
+    #Function is odd and divergent
+    assert summation(1 / k**3, (k, -oo, -1)) == -zeta(3)
```
Isn't this incorrect?

t-sibiraj replied:
The first assertion is incorrect and that test case case is not handled by my code so I will be removing that


But the second one is correct
![image](https://github.com/user-attachments/assets/6de58cbb-11ec-484b-bf2f-0e73bd90b02a)

t-sibiraj replied:
The first assertion is incorrect and that test case case is not handled by my code or the `eval_sum_resiude` function so I will be removing that(I also used breakpoints to confirm that the test case was not matched)


But the second one is correct
![image](https://github.com/user-attachments/assets/6de58cbb-11ec-484b-bf2f-0e73bd90b02a)

t-sibiraj replied:
I just noticed that. I used `summation` because the original GitHub issue referenced it, but I’ll change it back to `eval_sum_residue` where applicable and remove any test cases which I added that wouldn’t call `eval_sum_residue`

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1452,13 +1452,23 @@ def eval_sum_residue(f, i_a_b):
 
-    def is_even_function(numer, denom):
-        """Test if the rational function is an even function"""
+    def get_function_symmetery(numer, denom):
+        """
+        Returns 'even', 'odd', or 'neither' based on the symmetry of a
+        rational function.
+        """
```
Why is this needed?

Does the previous code not handle odd functions?

t-sibiraj replied:
Yes previously my code included custom logic to determine if a function was odd. However I realized that `is_even_function` already checks whether the numerator and denominator are even or odd. Since that logic is already in place, I expanded on it to compute the function’s symmetry instead of using custom logic.

t-sibiraj replied:
Previously, my code included custom logic to determine if a function was odd. However, I realized that is_even_function already checks whether the numerator and denominator are even or odd. Since that logic is already in place, I expanded on it to compute the function’s symmetry instead of using custom logic.

oscarbenjamin replied:
Are the formulae used by `eval_sum_residue` valid for odd functions? I'm wondering why it does not currently handle that case.

t-sibiraj replied:
What I found out is that the `eval_sum_residue` function is not called for the assertion case below I also verified that by setting break points and also by placing print statements inside the function

`assert summation((k), (k, -oo, 0)) == oo`

t-sibiraj committed:
fixed odd function logic

t-sibiraj committed:
remove unnecessary test case

oscarbenjamin commented:
I'm a bit confused here. I don't know the `eval_sum_residue` function very well but you have made some changes to it and I would like you to explain what the changes are.

You have added a case to handle when the summand is an odd function. It looks like previously `eval_sum_residue` did not handle the odd case and I don't know why that would be. Are the general formulae used by `eval_sum_residue`? Is there some other way that it handled odd functions before? What are examples where this now correctly handles odd summands?

Please don't assume that I know what all the code does and that you don't need to explain these things.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1535,10 +1545,15 @@ def get_residue_factor(numer, denom, alternating):
 
-    if not (a.is_finite and b is S.Infinity):
-        return None
+    if not ('even' == get_function_symmetery(numer, denom)):
+        #for odd function flip the limit and negate
+        #limit (-oo, a) is flipped to (-a, oo) and the answer is negated
+        if ('odd' == get_function_symmetery(numer, denom)):
```
The `get_function_symmetry` function should not be called more than once. Also the word symmetry is mispelt.

t-sibiraj committed:
correct spelling of the word symmetry and remove unnecessary function call

t-sibiraj committed:
correct spelling of the word rational

t-sibiraj committed:
remove unnecessary test cases which doesn't call eval_sum_residue

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1552,14 +1568,22 @@ def get_residue_factor(numer, denom, alternating):
         denom = denom.shift(shift)
 
-        if not is_even_function(numer, denom):
+        if not ('even' == get_function_symmetry(numer, denom)):
```
This is calling the function again.

Also these conditions read strangely with
```
if 'even' == ...:
```
It is usually more natural to write
```
if ... == 'even':
```

t-sibiraj replied:
I will update it to `if .... = 'even'` for better readability. The reason why I am calling it again is because the numer and denominator get's modified in the previous line and this was also the intended behaviour in the original code before I made any changes

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1584,6 +1584,9 @@ def test_summation_by_residues():
     assert eval_sum_residue((-1)**x / x**2, (x, S(1), oo)) == -pi**2/12
     assert eval_sum_residue((-1)**x / x**2, (x, S(2), oo)) == 1 - pi**2/12
 
+    # https://github.com/sympy/sympy/issues/27824
+    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == Rational(1, 2) + pi/(2*tanh(pi))
```
There are no tests here for odd functions. Can you give an example of an odd summand that is now handled that is not otherwise handled on master?

t-sibiraj replied:
I think adding an example of an odd summand that will be handled by  the `eval_sum_residue` is not possible because the `eval_sum_residue` tries to transform any function that is not `even` to `even` by shifting but since the code is classifies functions as odd or even based on the coefficients of the numerator and denominator, the calculated shift will be always zero for odd functions.

Example: For the odd function `1/k**5 + k**3,` the coefficient of the second highest power in denominator`(k**4)` is zero so the computed shift will be zero

https://github.com/sympy/sympy/blob/7cada5cc7423a3e2e95f6465747d33b86eb12cb5/sympy/concrete/summations.py#L1545-L1550

From the above code you can see when the shift = 0, the odd function doesn't proceed through the shifting making the `eval_sum_residue` to return `None`


When users wants to calculate odd summand or summands in general they will use the `summations` function which tries various strategies  in addition to eval_sum_residue that will potentially handle this behaviour otherwise returns None


```python
>>> summation(expr, (k, -oo, S(-1)))
-zeta(3)
```

The above example will get matched by one of the functions inside `summations.py` before it evens reaches the `eval_sum_residue`


So the odd function check is redundant in practise but mathematically a valid fallback case(Let me know if you still you'd still prefer to keep it in the logic for completeness)

oscarbenjamin replied:
I think it is fine for `eval_sum_residue` to handle some cases like odd functions even if it will not be used by `summation` for those cases. There needs to be explicit tests for `eval_sum_residue` with odd functions though.

t-sibiraj committed:
improved code readability

t-sibiraj commented:
I suppose [test / code-quality (pull_request)](https://gitub.com/sympy/sympy/actions/runs/14154391175/job/39651743753?pr=27827) is not failing because of the changes I made in the commit as I only re ordered the conditions

I also ran the code quality checks locally and they pass on my end

```bash
(venv) sibi@sibis-MBP sympy % ./bin/test code_quality                                
===================================================================== test session starts ======================================================================
platform darwin -- Python 3.11.9, pytest-8.3.5, pluggy-1.5.0
architecture: 64-bit
cache:        yes
ground types: python 

rootdir: /Users/sibi/Downloads/open_source/sympy
configfile: pyproject.toml
plugins: hypothesis-6.130.2
collected 7 items                                                                                                                                              

sympy/testing/tests/test_code_quality.py .......                                                                                                         [100%]

====================================================================== 7 passed in 6.30s =======================================================================
```

oscarbenjamin commented:
I think that a new version of flake8 has been released yesterday. Probably if you update to the latest version you will see the error.

t-sibiraj commented:
> I think that a new version of flake8 has been released yesterday. Probably if you update to the latest version you will see the error.

I merged your pull request into my branch and all my issues are resolved.

t-sibiraj committed:
add test cases for odd function for eval_sum_residue

t-sibiraj committed:
Merge remote-tracking branch 'upstream/master' into support-residue-sum-neg-inf-to-zero

t-sibiraj commented:
> > I merged your pull request into my branch and all my issues are resolved.
> 
> It is better not to do this. That other pull request is merged now meaning that the problem is fixed but now you have redundant commits here and there is a merge conflict.

I will fix this and keep this in mind next time

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1583,6 +1583,17 @@ def test_summation_by_residues():
+    # https://github.com/sympy/sympy/issues/27824
+    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == Rational(1, 2) + pi/(2*tanh(pi))
+
+    # The odd function when called directly in eval_sum_residue returns None
+    # because the current implementation attempts to convert non-even functions
+    # into even ones by shifting. However, this doesn't work in this case,
+    # as the coefficient of the second-highest power in the denominator (k**2)
+    # is zero. This results in a shift of zero, which means the function should
+    # return None.
+    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
```
Is there a test case where it does work?

t-sibiraj replied:
I don’t think so but I can add test cases of functions which are neither odd nor even that would work

oscarbenjamin replied:
Why does the code handle odd functions if there are no examples where it works.

t-sibiraj replied:
I mean as I mentioned earlier the odd function check is redundant in practise but mathematically a valid fallback case so you suggested to keep that logic in place even though it will not be used by the `summation` function. 

> I think it is fine for eval_sum_residue to handle some cases like odd functions even if it will not be used by summation for those cases.


So should I remove the odd function logic and remove the related test case or leave it as it is

t-sibiraj replied:
So you want me to remove the test case and the logic right since the summation function already has code to handle that

oscarbenjamin replied:
As I said before I think it is fine if `eval_sum_residue` can handle odd functions even though `summation` might not call it in that case. I don't understand why there can't be a test case for when `eval_sum_residue` does handle an odd function.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1535,10 +1545,18 @@ def get_residue_factor(numer, denom, alternating):
-    if not (a.is_finite and b is S.Infinity):
-        return None
+    rational_function_symmetry = get_function_symmetry(numer, denom)
+    if not (rational_function_symmetry == 'even'):
```
Rather than `if not (a == b)` you can write `if a != b`.

Actually though since inside here you alread have an `if rational_function_symmetry == 'odd'` it would be better to just use `if/elif/else`.

Harikrishna-Srinivasan commented:
Shouldn't `summation(1/(k**2+1), (k, 0, -oo))` be _-pi/(2*tanh(pi)) - 1/2_?

But with this PR, I got _summation(1/(k**2+1), (k, 0, -oo))_ as an unevaluated `sum`, that is _summation(1/(k**2+1), (k, 0, -oo))_. Should this be changed? (Isn't this an unexpected behaviour?)

t-sibiraj commented on sympy/concrete/summations.py:
```diff
@@ -1552,14 +1570,22 @@ def get_residue_factor(numer, denom, alternating):
         denom = denom.shift(shift)
 
-        if not is_even_function(numer, denom):
+        if not (get_function_symmetry(numer, denom) == 'even'):
```
The reason why I am calling it again is because the numerator and denominator get's modified in the previous line so we need to call the function again with the updated values and this was also the intended behaviour in the original code before I made any changes

t-sibiraj commented:
> Shouldn't `summation(1/(k**2+1), (k, 0, -oo))` be _-pi/(2*tanh(pi)) + 1/2_?
> 
> But with this PR, I got _summation(1/(k**2+1), (k, 0, -oo))_ as an unevaluated `sum`, that is _summation(1/(k**2+1), (k, 0, -oo))_. Should this be changed? (Isn't this an unexpected behaviour?)

Sympy returns the summation `(1/(k**2+1), (k, 0, -oo))` as unevaluated even before my PR. My PR only handles sums from -oo to a finite number and it doesn't affect ranges like 0 to -oo.


Before my PR:
```python3
>>> summation(1/(k**2+1), (k, 0, -oo))
Sum(1/(k**2 + 1), (k, 0, -oo))
```

With my PR(still not merged):
```python3
>>> summation(1/(k**2+1), (k, 0, -oo))
Sum(1/(k**2 + 1), (k, 0, -oo))
```

t-sibiraj committed:
refactor symmetry handling

oscarbenjamin commented:
This has merge conflicts now after gh-27893 was merged.

t-sibiraj committed:
Merge remote-tracking branch 'upstream/master' into support-residue-sum-neg-inf-to-zero

t-sibiraj committed:
resolve merge conflict

t-sibiraj committed:
added test cases for eval_sum_residue

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1603,28 @@ def test_summation_by_residues():
     assert eval_sum_residue((-1)**x / x**2, (x, S(2), oo)) == 1 - pi**2/12
 
+    # https://github.com/sympy/sympy/issues/27824
+    #An example of even function which works for -oo to k
+    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == Rational(1, 2) + pi/(2*tanh(pi))
+
+    #An example of function which is neither or odd which works for -oo to k
+    assert eval_sum_residue(1 / (k**2 + 1), (k, -S.Infinity, S(0))) == Rational(1,2) + pi/(2*tanh(pi))
```
These two test cases are identical even though the comments claim otherwise.

t-sibiraj replied:
I will change it to the correct one

t-sibiraj commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1603,28 @@ def test_summation_by_residues():
+
+    # An example of odd function which works for eval_sum_residue
+    assert simplify(eval_sum_residue(k / (k**4 + 1), (k, -S.Infinity, S.Infinity))) == 0
+    assert simplify(eval_sum_residue(k / (k**2 + 1)**2, (k, -S.Infinity, S.Infinity))) == 0
```
You can correct me if I am wrong but this is why I think both of them are odd functions

### For the first odd function

```python3
>>> numer
Poly(k, k, domain='ZZ')
>>> numer.monoms()
[(1,)]

>>> denom 
Poly(k**4 + 1, k, domain='ZZ')
>>> denom.monoms()
[(4,), (0,)]

>>> numer_odd = all(i % 2 == 1 for (i,) in numer.monoms())
>>> denom_even = all(i % 2 == 0 for (i,) in denom.monoms())
>>> numer_odd and denom_even
True
```

(Or)

```python3
>>> f = k / (k**4 + 1)
>>> f_negative = f.subs(k,-k)
>>> f + f_negative == 0
True
```


### For the second function


```python3
>>> numer
Poly(k, k, domain='ZZ')
>>> numer.monoms()
[(1,)]

>>> denom
Poly(k**4 + 2*k**2 + 1, k, domain='ZZ')
>>> denom.monoms()
[(4,), (2,), (0,)]

>>> numer_odd = all(i % 2 == 1 for (i,) in numer.monoms())
>>> denom_even = all(i % 2 == 0 for (i,) in denom.monoms())
>>> numer_odd and denom_even
True
```

(Or)

```python3
>>> f = k / (k**2 + 1)**2
>>> f_negative = f.subs(k, -k)
>>> f + f_negative == 0
True
```

oscarbenjamin replied:
Oh sorry they are odd functions. I missed the `k` in the numerator.

Can we test a case where the new code added here would apply and the answer is not just zero? Summing odd functions is trivial if the limits are `(-oo, oo)`.

smichr replied:
Is it impossible to get an odd case because when you convert it to an even case you do so by shifting and removing one of the possibly symmetric roots needed to pass:

```python
    poles = get_poles(denom)
    if poles is None:
        return None
    int_roots, nonint_roots = poles

    if int_roots:
        int_roots = [int(root) for root in int_roots]
        int_roots_max = max(int_roots)
        int_roots_min = min(int_roots)
        # Integer valued poles must be next to each other
        # and also symmetric from origin (Because the function is even)
        if not len(int_roots) == int_roots_max - int_roots_min + 1:
            return None

        # Check whether the summation indices contain poles
        if a <= max(int_roots):
            return None
```

smichr replied:
And both of these tests already pass in master. e.g. `simplify(eval_sum_residue(k / (k**2 + 1)**2, (k, -S.Infinity, S.Infinity))) -> 0`

t-sibiraj committed:
fixed the neither or odd even function test case for eval_sum_residue

t-sibiraj committed:
Merge branch 'master' into support-residue-sum-neg-inf-to-zero

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1603,27 @@ def test_summation_by_residues():
+    #An example of even function which works for -oo to k
+    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == Rational(1, 2) + pi/(2*tanh(pi))
+
+    #An example of function which is neither or odd which works for -oo to k
```
```suggestion
    #An example of function which is neither even nor odd which works for -oo to k
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1603,27 @@ def test_summation_by_residues():
+    # is zero. This results in a shift of zero, which means the function should
+    # return None.
+    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
+
+    # An example of odd function which works for eval_sum_residue
+    assert simplify(eval_sum_residue(k / (k**4 + 1), (k, -S.Infinity, S.Infinity))) == 0
+    assert simplify(eval_sum_residue(k / (k**2 + 1)**2, (k, -S.Infinity, S.Infinity))) == 0
```
```suggestion
```

smichr committed:
Apply suggestions from code review

smichr committed:
simplify handling of the (-oo,k) limit

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1603,24 @@ def test_summation_by_residues():
+    # as the coefficient of the second-highest power in the denominator (k**2)
+    # is zero. This results in a shift of zero, which means the function should
+    # return None.
+    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
```
```suggestion
    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
    # SO issue cited on #27827
    assert summation(1 / ((k+1) ** 4 + 1), (k, -oo, oo)).simplify() == sqrt(2)*pi*(1 - I)*(I/tan(sqrt(2)*pi*(1 + I)/2) + 1/tan(sqrt(2)*pi*(1 - I)/2))/4
```

smichr committed:
test for function that can be made even with inf limits

smichr committed:
handle func made even having inf limits -oo,oo

smichr committed:
Update summations.py

smichr committed:
Update test_sums_products.py

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -26,6 +26,7 @@
 from sympy.functions.special.gamma_functions import (gamma, lowergamma)
 from sympy.functions.special.tensor_functions import KroneckerDelta
 from sympy.functions.special.zeta_functions import zeta
+from sympy.functions.elementary.trigonometric import tab
```
```suggestion
from sympy.functions.elementary.trigonometric import tan
```

smichr committed:
Update sympy/concrete/tests/test_sums_products.py

smichr commented:
I am also handling the non-integer shift that is legal for [-oo,oo] sums if it can make them even. And a test using the example from the SO case given on the issue page was added.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1530,56 +1530,53 @@ def get_residue_factor(numer, denom, alternating):
+            return eval_sum_residue(f, (i, a-shift, b-shift))
+
+        elif a.is_Integer or b.is_Integer:
             return None
```
The confusing thing here is that the logic is in the wrong order. It should be checking for the `(-oo, oo)` case first. If both limits are infinite we don't care if the summand is even so it should not even check `is_even_function`. Instead that case should be handled up front:
```python
    if (a, b) == (S.NegativeInfinity, S.Infinity):

        poles = get_poles(denom)
        if poles is None:
            return None
        int_roots, nonint_roots = poles
        if int_roots:
            return None

        residue_factor = get_residue_factor(numer, denom, alternating)
        residues = [residue(residue_factor, z, root) for root in nonint_roots]
        return -S.Pi * sum(residues)

    if a is S.NegativeInfinity:
        a, b = -b, -a

    assert a.is_Integer and b is S.Infinity

    if not is_even_function(numer, denom):
        # Try shifting summation and check if the summand can be made
        # even wrt the origin.
        # Sum(f(n), (n, a, b)) => Sum(f(n + s), (n, a - s, b - s))
        shift = get_shift(denom)

        if not shift.is_Integer:
            return None
        if shift == 0:
            return None

        numer = numer.shift(shift)
        denom = denom.shift(shift)

        if not is_even_function(numer, denom):
            return None
```

smichr replied:
As noted in the test for the SO issue -- the -oo,oo odd function fails if it is not made even.

oscarbenjamin replied:
I don't see anything that verifies that it is odd.

What are the conditions for the formulae in the docstring to be correct? If they require the summand to be even or odd then I think that should be stated there.

oscarbenjamin replied:
I think that the SO issue is actually a bug in `residue` that it doesn't work with RootOf:
```python
(Pdb) p root
CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)
(Pdb) p residue_factor
cot(_z*pi)/(_z**4 + 4*_z**3 + 6*_z**2 + 4*_z + 2)
(Pdb) p residue(residue_factor, z, root)
0
(Pdb) p residue(residue_factor.factor(extension=root), z, root)
1/((2 + 2*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0))*(-3*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)**2 - 2*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0) - CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)**3)*(2 + CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)**3 + 4*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0) + 3*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)**2)*tan(pi*CRootOf(x**4 + 4*x**3 + 6*x**2 + 4*x + 2, 0)))
(Pdb) p residue(residue_factor.factor(extension=root), z, root).evalf()
-0.171645037817808 + 0.179610797637112*I
```

smichr replied:
I still get 0 and it takes a long time. I made the change, however.

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1604,26 @@ def test_summation_by_residues():
+
+    #An example of function which is neither even nor odd which works for -oo to k
+    assert eval_sum_residue(1 / (k**2 + 2*k +2), (k, -oo, S(0))) == 1 + pi/(2*tanh(pi))
+    assert eval_sum_residue(1 / (k**2 + 2*k +2), (k, -oo, -S(1))) == S.Half+pi/tanh(pi)/2
```
```suggestion
    ans = Rational(1, 2) + pi/(2*tanh(pi))
    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == ans
    assert eval_sum_residue((1/(k**2+1)), (k, oo, -S(1))) == -ans  # Karr convention
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1604,26 @@ def test_summation_by_residues():
+    # is zero. This results in a shift of zero, which means the function should
+    # return None.
+    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
+    # SO issue cited on #27827
+    assert summation(1 / ((k+1) ** 4 + 1), (k, -oo, oo)).simplify() == sqrt(2)*pi*(1 - I)*(I/tan(sqrt(2)*pi*(1 + I)/2) + 1/tan(sqrt(2)*pi*(1 - I)/2))/4
```
```suggestion
    # SO issue cited on #27827
    # this will return 0 if it is not allowed to shift to be even
    assert summation(1 / ((k + 1) ** 4 + 1), (k, -oo, oo)).simplify() == (sqrt(2)*pi*(1 - I)*(I/tan(sqrt(2)*pi*(1 + I)/2) +
        1/tan(sqrt(2)*pi*(1 - I)/2))/4)
```

smichr committed:
Apply suggestions from code review

smichr committed:
negate result when a>b per Karr convention

smichr commented on sympy/concrete/summations.py:
```diff
@@ -1530,56 +1530,53 @@ def get_residue_factor(numer, denom, alternating):
-        else:
-            f = numer.as_expr() / denom.as_expr()
-        return eval_sum_residue(f, (i, a-shift, b-shift))
+    if a is S.NegativeInfinity and b is not S.Infinity:
+        a, b = -b, -a
```
```suggestion
```

smichr commented on sympy/concrete/summations.py:
```diff
@@ -1530,56 +1530,53 @@ def get_residue_factor(numer, denom, alternating):
+        residues = [residue(residue_factor, z, root) for root in nonint_roots]
+        return -S.Pi * sum(residues)
+
+    assert a.is_Integer and b is S.Infinity
```
```suggestion
    if a is S.NegativeInfinity:
        a, b = -b, -a
    assert a.is_Integer and b is S.Infinity  # and the function is even
```

smichr committed:
Apply suggestions from code review

smichr committed:
extension=True when handling residues

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1491,29 +1504,14 @@ def get_shift(denom):
-    # Quick exit heuristic for the sums which doesn't have infinite range
-    if a != S.NegativeInfinity and b != S.Infinity:
-        return None
+        return [residue(residue_factor.factor(extension=True), z, root) for root in roots]
```
It should be extension=root rather than extension=True. I showed the wrong thing before.

smichr replied:
Feel free to modify residue calculations. All tests are now passing locally.

smichr committed:
allow (-oo,oo) to continue with non-integer shift

smichr committed:
Update summations.py

smichr committed:
Update summations.py

smichr committed:
Update sympy/concrete/summations.py

oscarbenjamin committed:
perf(concrete): add _get_residue for eval_sum_residue

The _get_residue function gets the residue but handles the case of an
algebraic root better than the general residue function. We can take
advantage of the specific structure of the summands handled by
eval_sum_residue rather than needing to factorise in the extension
field.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1530,66 +1521,70 @@ def get_residue_factor(numer, denom, alternating):
+                f = numer.as_expr() / denom.as_expr()
+            return eval_sum_residue(f, (i, a-shift, b-shift))
+        elif not both_inf:
+            return None  # can't make a shift that keeps limits integer
 
-        if not is_even_function(numer, denom):
+    if both_inf:  # even or not
```
I think that the both_inf case can be handled above before checking `is_even_function` now that the residue calculation is fixed.

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1603,6 +1604,30 @@ def test_summation_by_residues():
+    assert eval_sum_residue((1/(k**2+1)), (k, -oo, S(0))) == ans
+    assert eval_sum_residue((1/(k**2+1)), (k, oo, -S(1))) == -ans  # Karr convention
+
+    # Odd functions over (-oo to k) are handled by other parts of the summation
+    # logic, so eval_sum_residue will return None in such cases.
+    # The odd function when called directly in eval_sum_residue returns None
+    # because the current implementation attempts to convert non-even functions
+    # into even ones by shifting. However, this doesn't work in this case,
+    # as the coefficient of the second-highest power in the denominator (k**2)
+    # is zero. This results in a shift of zero, which means the function should
+    # return None.
+    assert eval_sum_residue((1/k**3), (k,-oo, S(-1))) == None
+    # SO issue cited on #27827
+    # this will return 0 if it is not allowed to shift to become even
```
Are these comments still true?

I don't think we need the lengthy first comment in any case.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1613,6 +1608,57 @@ def get_residue_factor(numer, denom, alternating):
     return result
 
 
+def _get_residue(numer: Poly, denom: Poly, a: Expr, alternating: bool) -> Expr:
+    """``Res((cot/csc)(pi*z)*numer(z)/denom(z))`` at ``a``."""
+    #
+    # Helper for eval_sum_residue. Since we know that a is a root of denom, we
+    # can factor out ``(x - a)^p`` from denom exactly using polynomials over
+    # the extension field containing ``a``. This is more efficient and more
+    # robust than the current implementation of ``residue`` which uses a series
+    # expansion.
```
It probably makes sense to move some of this to the residue function although there is a long comment there explaining why a different approach should be used anyway.

smichr replied:
These changes fail locally for me:

```
================================== FAILURES ===================================
_________________________ test_summation_by_residues __________________________

    def test_summation_by_residues():
        x = Symbol('x')
    
        # Examples from Nakhle H. Asmar, Loukas Grafakos,
        # Complex Analysis with Applications
        assert eval_sum_residue(1 / (x**2 + 1), (x, -oo, oo)) == pi/tanh(pi)
        assert eval_sum_residue(1 / x**6, (x, S(1), oo)) == pi**6/945
        assert eval_sum_residue(1 / (x**2 + 9), (x, -oo, oo)) == pi/(3*tanh(3*pi))
        assert eval_sum_residue(1 / (x**2 + 1)**2, (x, -oo, oo)).cancel() == \
            (-pi**2*tanh(pi)**2 + pi*tanh(pi) + pi**2)/(2*tanh(pi)**2)
        assert eval_sum_residue(x**2 / (x**2 + 1)**2, (x, -oo, oo)).cancel() == \
            (-pi**2 + pi*tanh(pi) + pi**2*tanh(pi)**2)/(2*tanh(pi)**2)
        assert eval_sum_residue(1 / (4*x**2 - 1), (x, -oo, oo)) == 0
        assert eval_sum_residue(x**2 / (x**2 - S(1)/4)**2, (x, -oo, oo)) == pi**2/2
        assert eval_sum_residue(1 / (4*x**2 - 1)**2, (x, -oo, oo)) == pi**2/8
>       assert eval_sum_residue(1 / ((x - S(1)/2)**2 + 1), (x, -oo, oo)) == pi*tanh(pi)
E       assert pi/coth(pi) == (pi * tanh(pi))
E        +  where pi/coth(pi) = eval_sum_residue((1 / (((x - (1 / 2)) ** 2) + 1)), (x, -oo, oo))
E        +    where 1 = S(1)
E        +  and   tanh(pi) = tanh(pi)

C:\Users\Admin\AppData\Local\Programs\Python\Python313\Lib\site-packages\sympy\concrete\tests\test_sums_products.py:1571: AssertionError
                               DO *NOT* COMMIT!                               
=========================== short test summary info ===========================
FAILED ..\..\Users\Admin\AppData\Local\Programs\Python\Python313\Lib\site-packages\sympy\concrete\tests\test_sums_products.py::test_summation_by_residues
=========== 1 failed, 88 passed, 3 deselected, 2 xfailed in 41.45s ============
<ExitCode.TESTS_FAILED: 1>
```

smichr committed:
Update summations.py

smichr committed:
Update test_sums_products.py

smichr committed:
Update test_sums_products.py
